pkg/cli/image/strategy/onerror.go

```diff
@@ -162,3 +154,150 @@ func readICSPsFromFile(icspFile string) ([]operatorv1alpha1.ImageContentSourcePo
+}
+
+var _ registryclient.AlternateBlobSourceStrategy = &onErrorIDMSStrategy{}
+
+// NewIDMSOnErrorStrategy returns IDMS alternate strategy which reads alternate
+// sources only after getting an error from the original requested.
+func NewIDMSOnErrorStrategy(file string) registryclient.AlternateBlobSourceStrategy {
+	return &onErrorIDMSStrategy{
+		idmsFile:              file,
+		alternates:            make(map[reference.DockerImageReference][]reference.DockerImageReference),
+		readIDMSsFromFileFunc: readIDMSsFromFile,
+	}
+}
+
+func (s *onErrorIDMSStrategy) FirstRequest(ctx context.Context, locator reference.DockerImageReference) (alternateRepositories []reference.DockerImageReference, err error) {
+	return nil, nil
+}
+
+func (s *onErrorIDMSStrategy) OnFailure(ctx context.Context, locator reference.DockerImageReference) (alternateRepositories []reference.DockerImageReference, err error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if alternates, ok := s.alternates[locator]; ok {
+		return alternates, nil
+	}
+	alternates, err := s.resolve(ctx, locator)
+	if err != nil {
+		return nil, err
+	}
+	if len(alternates) == 0 {
+		return nil, fmt.Errorf("no alternative image references found for image: %s", locator.String())
+	}
+	s.alternates[locator] = alternates
+	return s.alternates[locator], nil
+}
+
+// resolve gathers possible image sources for a given image
+// gathered from ImageContentSourcePolicy file.
+// Image reference of user-given image may be different from original in case of mirrored images.
+func (s *onErrorIDMSStrategy) resolve(ctx context.Context, imageRef reference.DockerImageReference) ([]reference.DockerImageReference, error) {
+	if len(s.idmsFile) == 0 {
+		return nil, fmt.Errorf("no ImageDigestMirrorSet specified")
+	}
+	klog.V(5).Infof("Reading IDMS from file %s", s.idmsFile)
+	idmsList, err := s.readIDMSsFromFileFunc(s.idmsFile)
+	if err != nil {
+		return nil, err
+	}
+	// always add the original as the first reference
+	imageRefList, err := alternativeImageSourcesIDMS(imageRef, idmsList, false)
+	if err != nil {
+		return nil, err
+	}
+	return imageRefList, nil
+}
+
+// alternativeImageSourcesIDMS returns unique list of DockerImageReference objects from list of ImageDigestMirrorSet objects
+// addSourceAsLastAlternate decides whether the original imageRef is first or the last element in the result
+func alternativeImageSourcesIDMS(imageRef reference.DockerImageReference, idmsList []apicfgv1.ImageDigestMirrorSet, addSourceAsLastAlternate bool) ([]reference.DockerImageReference, error) {
+	var imageSources []reference.DockerImageReference
+	klog.V(5).Infof("%v ImageReference added to potential ImageSourcePrefixes from ImageDigestMirrorSet", imageRef.AsRepository().AsV2())
+	if !addSourceAsLastAlternate {
+		imageSources = append(imageSources, imageRef.AsRepository().AsV2())
+	}
+	repo := imageRef.AsRepository().Exact()
+	for _, idms := range idmsList {
+		repoDigestMirrors := idms.Spec.ImageDigestMirrors
```

Comment:
This PR assumes that ICSP and IDMS have same functionality. If there is a behavioral difference between these two types, we need to make changes in here. For example, IDMS has `mirrorSourcePolicy` field and we ignores this field in here, is that correct @QiWang19 ?.

Reply:
[mirrorSourcePolicy](https://github.com/openshift/api/blob/f36cfb9a990998106c5ebaa4fad0c7ff1264a2f5/config/v1/0000_10_config-operator_01_imagedigestmirrorset.crd.yaml#L50) defines whether crio pulls the image from the source registry when pulling. I think
-  if the mirrorSourcePolicy is specified as `NeverContactSource`, the source should not add to `imageSources`, neither the first time nor the last item, so the oc behavior is compatible with the container runtime. 
- if `NeverContactSource`is not `NeverContactSource` we use the same logic as current `alternativeImageSources()`

WDYT?

Reply:
Sounds fine to me. But `idms.Spec.ImageDigestMirrors` is an array, what should be the behavior when one item in this array contains `NeverContactSource` and the others have `AllowContactingSource`?

Reply:
I think we can break the loop and error out. 
`NeverContactSource` and `AllowContactingSource` for the same source is conflicting. MCO returns the error when rolling out the IDMS CR.

Reply:
Thanks. I'll change that way.

Reply:
@QiWang19 updated. Could you PTAL?